### PR TITLE
new quickstart sample 201-device-provisioning-service

### PR DESCRIPTION
**Pull request by kgremban**

### CODE REVIEW

File: quickstart/201-iot-hub-with-device-provisioning-service/main.tf [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+}
+
+resource "azurerm_resource_group" "rg" {
+  location = var.resource_group_location
+  name     = random_pet.rg_name.id
+}
+
+# Create storage account & container
+resource "random_string" "sa_name" {
+  length = 12
```
Comment (lonegunmanb) [inline quickstart/201-iot-hub-with-device-provisioning-service/main.tf]: Could we run `terraform fmt` command for this directory to align these equal marks please?

File: quickstart/201-iot-hub-with-device-provisioning-service/main.tf [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+    }
+  }
+
+  tags = {
+    purpose = "testing"
+  }
+}
+
+#Create IoT Hub Access Policy
+resource "azurerm_iothub_shared_access_policy" "hubaccesspolicy" {
```
Comment (lonegunmanb) [inline quickstart/201-iot-hub-with-device-provisioning-service/main.tf]: Could we rename this resource to `hub_access_policy` to maintain a consist naming convention for this module?

File: quickstart/201-iot-hub-with-device-provisioning-service/main.tf [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+
+resource "azurerm_storage_container" "my_terraform_container" {
+  name                  = "mycontainer"
+  storage_account_name  = azurerm_storage_account.sa.name
+  container_access_type = "private"
+}
+
+
+# Create an Event Hub & Authorization Rule
+resource "random_pet" "eventhubnamespace_name" {
```
Comment (lonegunmanb) [inline quickstart/201-iot-hub-with-device-provisioning-service/main.tf]: Could we rename this resource to `eventhub_namespace_name` to maintain a consist naming convention for this module?

File: quickstart/201-iot-hub-with-device-provisioning-service/main.tf [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+  prefix = var.eventhub_namespace_name_prefix
+}
+
+resource "azurerm_eventhub_namespace" "namespace" {
+  name                = random_pet.eventhubnamespace_name.id
```
Comment (lonegunmanb) [inline quickstart/201-iot-hub-with-device-provisioning-service/main.tf]: Would `eventhub_namespace_name` for `random_pet`'s name be better here?

File: quickstart/201-iot-hub-with-device-provisioning-service/variables.tf [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+  default     = "eastus"
+  description = "Location of the resource group."
+}
+
+variable "resource_group_name_prefix" {
+  default     = "rg"
+  description = "Prefix of the resource group name that's combined with a random ID so name is unique in your Azure subscription."
+}
+
+variable "storage_account_name_prefix" {
```
Comment (lonegunmanb) [inline quickstart/201-iot-hub-with-device-provisioning-service/variables.tf]: This variable is an unused variable, could we just remove it?

File: quickstart/201-iot-hub-with-device-provisioning-service/providers.tf [hunk 6]
```diff
@@ -0,0 +1,18 @@
+terraform {
+  required_version = ">=0.12"
+
+  required_providers {
+    azurerm = {
+      source  = "hashicorp/azurerm"
+      version = "~>2.0"
```
Comment (lonegunmanb) [inline quickstart/201-iot-hub-with-device-provisioning-service/providers.tf]: The mainstream version of `azurerm` provider is 3.x now and 2.x is no longer maintained anymore, would the following restriction be better?:

```hcl
version = ">=3.0"
```

File: quickstart/201-iot-hub-with-device-provisioning-service/providers.tf [hunk 7]
```diff
@@ -0,0 +1,18 @@
+terraform {
+  required_version = ">=0.12"
```
Comment (lonegunmanb) [inline quickstart/201-iot-hub-with-device-provisioning-service/providers.tf]: The mainstream version of Terraform is 1.3 now, could we adjust this `required_version` to `>=1.0`?

**Review by lonegunmanb — CHANGES_REQUESTED** [review]
Thanks @kgremban for opening this pr! Almost LGTM but some tiny issues need to be taken care of. Appreciate for your understanding.

File: quickstart/201-iot-hub-with-device-provisioning-service/main.tf (hunk 4 above)
Comment (kgremban) [reply quickstart/201-iot-hub-with-device-provisioning-service/main.tf]: @lonegunmanb I'm not sure what this means. Could you show me what the updated line of code would look like?

**Review by grayzu — APPROVED** [review]
Thanks for the PR! This looks great to me.

File: quickstart/201-iot-hub-with-device-provisioning-service/main.tf [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+}
+
+resource "azurerm_iothub" "iothub" {
+  name                = random_pet.iothub_name.id
+  resource_group_name = azurerm_resource_group.rg.name
+  location            = azurerm_resource_group.rg.location
+
+  sku {
+    name     = "S1"
+    capacity = "1"
```
Comment (lonegunmanb) [inline quickstart/201-iot-hub-with-device-provisioning-service/main.tf]: According to the [provider's schema](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/internal/services/iothub/iothub_resource.go#L119-L122), the `capacity`'s type is `TypeInt`. Terraform will [auto convert](https://www.terraform.io/language/expressions/types#type-conversion) `string` to `number`, but could we change `"1"` to `1` to make type more precisely please? Thanks.

File: quickstart/201-iot-hub-with-device-provisioning-service/main.tf [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+
+#Create IoT Hub Access Policy
+resource "azurerm_iothub_shared_access_policy" "hub_access_policy" {
+  name                = "terraform-policy"
+  resource_group_name = azurerm_resource_group.rg.name
+  iothub_name         = azurerm_iothub.iothub.name
+
+  registry_read  = true
+  registry_write = true
+  service_connect = true
```
Comment (lonegunmanb) [inline quickstart/201-iot-hub-with-device-provisioning-service/main.tf:139]: Could we run `terraform fmt` for this folder to align these equal symbols? Thanks.

File: quickstart/201-iot-hub-with-device-provisioning-service/main.tf [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+
+resource "azurerm_iothub_dps" "dps" {
+  name                = random_pet.dps_name.id
+  resource_group_name = azurerm_resource_group.rg.name
+  location            = azurerm_resource_group.rg.location
+  allocation_policy   = "Hashed"
+
+  sku {
+    name     = "S1"
+    capacity = "1"
```
Comment (lonegunmanb) [inline quickstart/201-iot-hub-with-device-provisioning-service/main.tf]: We have the same issue here according to the [provider's schema](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/internal/services/iothub/iothub_dps_resource.go#L75-L79).

**Review by lonegunmanb — CHANGES_REQUESTED** [review]
Thanks @kgremban for the update, almost LGTM but still some tiny issues, would you please take a look?

File: quickstart/201-iot-hub-with-device-provisioning-service/main.tf (hunk 4 above)
Comment (lonegunmanb) [reply quickstart/201-iot-hub-with-device-provisioning-service/main.tf]: Apology for the confusion. I saw you've renamed the `random_pet.eventhubnamespace_name` to `random_pet.eventhub_namespace_name` in line 38 and it looks good to me now. Thanks.

**Comment by kgremban** [comment]
Thanks @lonegunmanb. I've applied this second round of changes. 

Also to respond to your question from email, I started out with the sample named 101-etc but then decided it probably fit better in the 201-etc category because it deploys several interdependent services. I just updated the PR title to match.

**Comment by kgremban** [comment]
@lonegunmanb Can you review this PR again?

**Review by lonegunmanb — COMMENTED** [review]
LGTM @kgremban!

**Comment by kgremban** [comment]
@TomArcherMsft Do you need to review this PR before merging? Thanks!

**Comment by TomArcherMsft** [comment]
> @TomArcherMsft Do you need to review this PR before merging? Thanks!

Thanks for asking, @kgremban. I need to review all Terraform content. However, the code review is owned by @lonegunmanb and @grayzu, so if they're good with the code, I am as well :) Therefore, please feel free to sign off and @lonegunmanb will merge.

**Comment by kgremban** [comment]
Thanks everyone for your help getting this sample out!

#sign-off

**Pull request merged**